Source/WebKit/UIProcess/ProvisionalPageProxy.cpp

```diff
@@ -844,8 +854,12 @@ void ProvisionalPageProxy::didReceiveSyncMessage(IPC::Connection& connection, IP
     }
 
     RefPtr page = m_page.get();
-    if (page)
-        page->didReceiveSyncMessage(connection, decoder, replyEncoder);
+    if (page) {
+        if (decoder.messageReceiverName() == Messages::WebBackForwardList::messageReceiverName())
```

Comment:
I don't know if you'd consider it _cleaner_ but it would almost certainly be less code to put this check inside `WebPageProxy::didReceiveMessage` (and equivalent for sync messages) and choose to forward messages onto `WebBackForwardList` at that point. This I think would remove the need for some of these changes in `ProvisionalPageProxy.cpp` and `RemotePageProxy.cpp`.